### PR TITLE
Break read+write dependency in Stream copy op

### PR DESCRIPTION
**Pull request by benaadams**

### CODE REVIEW

File: src/Microsoft.AspNetCore.Http.Extensions/StreamCopyOperation.cs [hunk 1]
```diff
@@ -14,55 +14,86 @@ namespace Microsoft.AspNetCore.Http.Extensions
...
+            var buffer1 = ArrayPool<byte>.Shared.Rent(DefaultBufferSize);
+
+            Debug.Assert(buffer0 != null);
+            Debug.Assert(buffer1 != null);
+            
+            return ArrayPoolCopyToAsyncInternal(source, destination, buffer0, buffer1, length, cancellationToken);
+        }
 
-                while (true)
+        private static async Task ArrayPoolCopyToAsyncInternal(Stream source, Stream destination, byte[] buffer0, byte[] buffer1, long? length, CancellationToken cancellationToken)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Http.Extensions/StreamCopyOperation.cs]: `Internal` is redundant with `private`

File: src/Microsoft.AspNetCore.Http.Extensions/StreamCopyOperation.cs [hunk 2]
```diff
@@ -14,55 +14,86 @@ namespace Microsoft.AspNetCore.Http.Extensions
...
-                Debug.Assert(buffer != null);
+                return _completedTask;
+            }
+
+            var buffer0 = ArrayPool<byte>.Shared.Rent(DefaultBufferSize);
+            var buffer1 = ArrayPool<byte>.Shared.Rent(DefaultBufferSize);
+
+            Debug.Assert(buffer0 != null);
+            Debug.Assert(buffer1 != null);
+
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Http.Extensions/StreamCopyOperation.cs]: ThrowIfCancellationRequested

File: src/Microsoft.AspNetCore.Http.Extensions/StreamCopyOperation.cs [hunk 3]
```diff
@@ -14,55 +14,86 @@ namespace Microsoft.AspNetCore.Http.Extensions
...
-                    {
-                        return;
-                    }
+                    readLength = (int)Math.Min(length.Value, (long)readLength);
+                }
 
-                    cancel.ThrowIfCancellationRequested();
+                var lastReadTask = source.ReadAsync(buffer0, 0, readLength, cancellationToken);
+                var lastWriteTask = _completedTask;
+                var otherBuffer = true;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Http.Extensions/StreamCopyOperation.cs]: naming? swapBuffers?

**Comment by davidfowl** [comment]
After reading @stephentoub's comment on the corefx PR I'm not so sure this is a good idea.

File: src/Microsoft.AspNetCore.Http.Extensions/StreamCopyOperation.cs [hunk 4]
```diff
@@ -14,55 +14,86 @@ namespace Microsoft.AspNetCore.Http.Extensions
...
 
-                    int readLength = buffer.Length;
-                    if (bytesRemaining.HasValue)
+                while ((count = await lastReadTask) != 0)
+                {
+                    if (count == 0)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Http.Extensions/StreamCopyOperation.cs]: Didn't you just check this on line 58?

File: src/Microsoft.AspNetCore.Http.Extensions/StreamCopyOperation.cs [hunk 5]
```diff
@@ -14,55 +14,86 @@ namespace Microsoft.AspNetCore.Http.Extensions
...
                     {
-                        readLength = (int)Math.Min(bytesRemaining.Value, (long)readLength);
+                        // End of the source stream.
+                        break;
                     }
-                    int count = await source.ReadAsync(buffer, 0, readLength, cancel);
 
-                    if (bytesRemaining.HasValue)
+                    readLength = (otherBuffer ? buffer1 : buffer0).Length;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Http.Extensions/StreamCopyOperation.cs]: This might be less confusing if you just called them readBuffer, writeBuffer, holdingBuffer, etc. and swapped them as needed.

File: src/Microsoft.AspNetCore.Http.Extensions/StreamCopyOperation.cs [hunk 6]
```diff
@@ -14,55 +14,86 @@ namespace Microsoft.AspNetCore.Http.Extensions
...
+                    if (length.HasValue)
                     {
-                        bytesRemaining -= count;
+                        readLength = (int)Math.Min(length.Value, (long)readLength);
                     }
 
-                    // End of the source stream.
-                    if (count == 0)
+                    lastReadTask = source.ReadAsync((otherBuffer ? buffer1 : buffer0), 0, readLength, cancellationToken);
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Http.Extensions/StreamCopyOperation.cs]: One more ThrowIfCancellationRequested before ReadAsync.

**Comment by Tratcher** [comment]
It probably works but it's hard to follow.

**Comment by benaadams** [comment]
Comment @davidfowl is referring to https://github.com/dotnet/corefx/pull/5954#discussion_r52127427

**Comment by benaadams** [comment]
Worth making changes, or close :+1: :-1: ?

**Comment by benaadams** [comment]
meh
